File: app/src/main/java/com/bookcrossing/mobile/presenters/BookCreatePresenter.java
package com.bookcrossing.mobile.presenters;

import android.graphics.Bitmap;
import android.net.Uri;
import com.arellomobile.mvp.InjectViewState;
import com.bookcrossing.mobile.models.Book;
import com.bookcrossing.mobile.models.Date;
import com.bookcrossing.mobile.ui.create.BookCreateView;
import com.crashlytics.android.Crashlytics;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.storage.StorageMetadata;
import com.google.zxing.BarcodeFormat;
import com.google.zxing.EncodeHintType;
import com.google.zxing.MultiFormatWriter;
import com.google.zxing.WriterException;
import com.google.zxing.common.BitMatrix;
import com.miguelbcr.ui.rx_paparazzo2.entities.FileData;
import java.util.Calendar;
import java.util.EnumMap;
import java.util.Map;

import static android.graphics.Color.BLACK;
import static android.graphics.Color.WHITE;

@InjectViewState public class BookCreatePresenter extends BasePresenter<BookCreateView> {

  private Book book;
  private Uri tempCoverUri;

  public BookCreatePresenter() {
    super();
    this.book = new Book();
    book.setFree(true);
  }

  private void uploadCover(String key) {
    if (tempCoverUri != null && firebaseWrapper.getAuth().getCurrentUser() != null) {
      StorageMetadata metadata = new StorageMetadata.Builder().setContentType("image/jpeg").build();
      resolveCover(key).putFile(tempCoverUri, metadata);
    }
  }

  public void saveCoverTemporarily(FileData result) {
    tempCoverUri = Uri.fromFile(result.getFile());
    getViewState().onCoverChosen(tempCoverUri);
  }

  private Bitmap encodeBookAsQrCode(String contents, BarcodeFormat format, int img_width,
      int img_height) throws WriterException {
    if (contents == null) {
      return null;
    }
    Map<EncodeHintType, Object> hints = new EnumMap<>(EncodeHintType.class);
    hints.put(EncodeHintType.CHARACTER_SET, "UTF-8");
    MultiFormatWriter writer = new MultiFormatWriter();
    BitMatrix bitMatrix;
    try {
      bitMatrix = writer.encode(contents, format, img_width, img_height, hints);
    } catch (IllegalArgumentException iae) {
      return null;
    }
    int width = bitMatrix.getWidth();
    int height = bitMatrix.getHeight();
    int[] pixels = new int[width * height];
    for (int y = 0; y < height; y++) {
      int offset = y * width;
      for (int x = 0; x < width; x++) {
        pixels[offset + x] = bitMatrix.get(x, y) ? BLACK : WHITE;
      }
    }

    Bitmap bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
    bitmap.setPixels(pixels, 0, width, 0, 0, width, height);
    return bitmap;
  }

  public void onNameChange(String name) {
    book.setName(name);
    getViewState().onNameChange();
  }

  public void onAuthorChange(String author) {
    book.setAuthor(author);
  }

  public void onPositionChange(String position) {
    book.setPositionName(position);
  }

  public void onDescriptionChange(String description) {
    book.setDescription(description);
  }

  public void publishBook() {
    book.setCity(getCity());
    setPublicationDate();
    final DatabaseReference newBookReference = books().push();
    newBookReference.setValue(book).addOnSuccessListener(aVoid -> {
      String key = newBookReference.getKey();
      uploadCover(key);
      getViewState().onReleased(key);
    }).addOnFailureListener(e -> {
      Crashlytics.logException(e);
      getViewState().onFailedToRelease();
    });
  }

  private void setPublicationDate() {
    Calendar calendar = Calendar.getInstance();
    Date date = new Date(calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH),
        calendar.get(Calendar.DAY_OF_MONTH), new java.util.Date().getTime());
    book.setWentFreeAt(date);
  }

  public Bitmap generateQrCode(String key) {
    try {
      return encodeBookAsQrCode(buildBookUri(key).toString(), BarcodeFormat.QR_CODE, 450, 450);
    } catch (WriterException e) {
      e.printStackTrace();
      Crashlytics.logException(e);
      return null;
    }
  }
}

File: app/src/main/java/com/bookcrossing/mobile/presenters/BookCreatePresenter.kt
/*
 *    Copyright 2019 Andrey Mukamolov
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.bookcrossing.mobile.presenters

import android.content.ContentResolver
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import androidx.core.content.edit
import com.bookcrossing.mobile.code.BookStickerSaver
import com.bookcrossing.mobile.code.QrCodeEncoder
import com.bookcrossing.mobile.models.BookBuilder
import com.bookcrossing.mobile.models.Date
import com.bookcrossing.mobile.ui.create.BookCreateView
import com.bookcrossing.mobile.util.EXTRA_CITY
import com.bookcrossing.mobile.util.EXTRA_DEFAULT_CITY
import com.crashlytics.android.Crashlytics
import com.google.firebase.storage.StorageMetadata
import com.google.zxing.WriterException
import durdinapps.rxfirebase2.RxFirebaseAuth
import durdinapps.rxfirebase2.RxFirebaseDatabase
import durdinapps.rxfirebase2.RxFirebaseStorage
import io.reactivex.Maybe
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import moxy.InjectViewState
import timber.log.Timber
import java.io.File
import java.io.IOException
import java.util.Calendar
import java.util.UUID

/**
 * Presenter for book create view
 */
@InjectViewState
class BookCreatePresenter : BasePresenter<BookCreateView>() {

  private val book: BookBuilder = BookBuilder()
  private lateinit var tempCoverUri: Uri

  private fun uploadCover(key: String): Observable<String>? {
    val metadata = StorageMetadata.Builder()
      .setContentType("image/jpeg")
      .build()
    return RxFirebaseAuth.observeAuthState(firebaseWrapper.auth)
      .filter { auth -> auth.currentUser != null }
      .switchMapSingle {
        RxFirebaseStorage.putFile(resolveCover(key), tempCoverUri, metadata)
          .map { key }
      }
      .onErrorReturn { key }
  }

  fun saveCoverTemporarily(coverUri: Uri?) {
    if (coverUri != null) {
      tempCoverUri = coverUri
    }
    viewState.onCoverChosen(tempCoverUri)
  }

  fun compressCoverPhoto(contentResolver: ContentResolver) {
    var coverPhoto: Bitmap? = null
    contentResolver.openInputStream(tempCoverUri)
      .use {
        coverPhoto = BitmapFactory.decodeStream(it)
      }
    contentResolver.openOutputStream(tempCoverUri)
      .use {
        coverPhoto?.compress(Bitmap.CompressFormat.JPEG, 60, it)
      }
    viewState.onCoverChosen(tempCoverUri)
  }

  @Throws(IOException::class)
  fun createImageFile(storageDir: File?): File {
    return File.createTempFile(
      "cover_${UUID.randomUUID()}_",
      ".jpg",
      storageDir
    ).also {
      tempCoverUri = Uri.fromFile(it)
    }
  }

  fun onNameChange(name: String) {
    book.setName(name)
    viewState.showCover()
  }

  fun onAuthorChange(author: String) {
    book.setAuthor(author)
  }

  fun onPositionChange(position: String) {
    book.setPositionName(position)
  }

  fun onDescriptionChange(description: String) {
    book.setDescription(description)
  }

  fun publishBook(city: String): Observable<String> {
    setPublicationDate()
    val newBook = book.createBook()
    newBook.city = city
    val newBookReference = books().push()
    val key = newBookReference.key.orEmpty()

    return RxFirebaseDatabase.setValue(newBookReference, newBook)
      .andThen(uploadCover(key))
      .subscribeOn(Schedulers.io())
      .observeOn(AndroidSchedulers.mainThread())
      .doOnNext {
        viewState.onReleased(it)
      }
      .doOnError {
        Crashlytics.logException(it)
        Timber.e(it, "Failed to release book")
        viewState.onFailedToRelease()
      }
  }

  private fun setPublicationDate() {
    val calendar = Calendar.getInstance()
    val date = Date(
      calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH),
      calendar.get(Calendar.DAY_OF_MONTH), calendar.timeInMillis
    )
    book.setWentFreeAt(date)
  }

  fun generateQrCode(key: String): Bitmap? {
    return try {
      QrCodeEncoder().encode(buildBookUri(key).toString())
    } catch (e: WriterException) {
      Timber.e("Failed to encode book key to QR code")
      Crashlytics.logException(e)
      null
    } catch (e: IllegalArgumentException) {
      Timber.e("Failed to save QR code in bitmap")
      Crashlytics.logException(e)
      null
    }
  }

  /**
   * Save generated sticker to device's default location for pictures
   *
   * @param stickerName Image name
   * @param stickerDescription Image description saved to metadata
   * @param sticker Generated image of the sticker
   */
  fun saveSticker(
    sticker: Bitmap,
    stickerName: String,
    stickerDescription: String
  ) {
    BookStickerSaver(systemServicesWrapper.app.contentResolver).saveSticker(
      stickerName, stickerDescription, sticker
    )
  }

  /**
   * Determine the city in which the user currently resides to set location of the released book
   */
  fun resolveUserCity(): Maybe<String> {
    return systemServicesWrapper.locationRepository.getLastKnownUserLocation()
      .flatMapMaybe<String> { location ->
        systemServicesWrapper.locationRepository.resolveUserCity(
          location
        )
      }
      .doOnSuccess { city -> saveCity(city) }
      .doOnError {
        Timber.e(it, "Failed to resolve city")
        viewState.askUserToProvideDefaultCity()
      }
  }

  /**
   * Save city to preferences
   */
  fun saveCity(city: String) {
    systemServicesWrapper.preferences.edit {
      putString(EXTRA_CITY, city)
      putString(EXTRA_DEFAULT_CITY, city)
    }
  }
}
